### PR TITLE
highfive: backport and fixes.

### PR DESCRIPTION
**Pull request by 1uc**
Includes:
  * allowing HDF5 from config files.
  * It seems that in `depends_on(..., when=cond)` we need to cover all three options: `""`, `"~mpi"` and `"+mpi"`. However, we must not have `depends_on("hdf5 +mpi")` which causes a conflict that we want both `hdf5+mpi` and `hdf5~mpi`.
  * When installing `highfive@master{+,~}mpi` and `libsonata@0.1.17` I keep getting linking error related to `-lhdf5-shared`. They went away after setting `-DHIGHFIVE_USE_INSTALL_DEPS:Bool=Off`.

### CODE REVIEW

File: bluebrain/repo-patches/packages/highfive/package.py [hunk 1]
```diff
@@ -79,6 +82,5 @@ def cmake_args(self):
             "-DHIGHFIVE_EXAMPLES:Bool=" + str(self.spec.satisfies("@develop")),
             "-DHIGHFIVE_UNIT_TESTS:Bool=" + str(self.spec.satisfies("@develop")),
             "-DHIGHFIVE_TEST_SINGLE_INCLUDES:Bool=" + str(self.spec.satisfies("@develop")),
-            "-DHIGHFIVE_USE_INSTALL_DEPS:Bool=ON",
-            "-DHDF5_NO_FIND_PACKAGE_CONFIG_FILE:Bool=ON",
+            "-DHIGHFIVE_USE_INSTALL_DEPS:Bool=Off",
```
Comment (matz-e) [inline bluebrain/repo-patches/packages/highfive/package.py:85]: Spack is built on the premise that packages should use the dependencies they have been installed with.

**Review by matz-e — CHANGES_REQUESTED** [review]

File: bluebrain/repo-patches/packages/highfive/package.py (hunk 1 above)
Comment (1uc) [reply bluebrain/repo-patches/packages/highfive/package.py:85]: If I switch to `"-DHIGHFIVE_USE_INSTALL_DEPS:Bool=On"` I can't install `libsonata` and `highfive` at the same time, e.g.
```
spack env create hi5-libsonata
spack env activate -p hi5-libsonata

spack add highfive
spack add libsonata@0.1.17

spack concretize -f && spack install --jobs 8
```
Fails with
```
     90     /gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_externals/install_gcc-12.2.0-skylake/hpe-mpi-2.25.hmpt-mgws7w/bin/mpicxx -fPIC -O2 -g -DNDEBUG -sh
            ared -Wl,-soname,libsonata.so.0.1 -o libsonata.so.0.1.17 CMakeFiles/sonata_shared.dir/src/common.cpp.o CMakeFiles/sonata_shared.dir/src/config.cpp
            .o CMakeFiles/sonata_shared.dir/src/edge_index.cpp.o CMakeFiles/sonata_shared.dir/src/edges.cpp.o CMakeFiles/sonata_shared.dir/src/hdf5_mutex.cpp.
            o CMakeFiles/sonata_shared.dir/src/node_sets.cpp.o CMakeFiles/sonata_shared.dir/src/nodes.cpp.o CMakeFiles/sonata_shared.dir/src/population.cpp.o 
            CMakeFiles/sonata_shared.dir/src/report_reader.cpp.o CMakeFiles/sonata_shared.dir/src/utils.cpp.o CMakeFiles/sonata_shared.dir/src/version.cpp.o  
            -Wl,-rpath,::: -lhdf5-shared -lpthread -lmpi
  >> 91     /gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_compilers/install_gcc-4.8.5-haswell/binutils-2.37-l5h4u2/bin/ld: cannot find -lhdf5-shared
  >> 92     collect2: error: ld returned 1 exit status
  >> 93     make[2]: *** [libsonata.so.0.1.17] Error 1
```

Comment (matz-e) [reply bluebrain/repo-patches/packages/highfive/package.py:85]: With the current state I see:
```
/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/highfive-2.6.2-snai4p/share/HighFive/CMake/HighFiveTargets.cmake
```
With:
```
set_target_properties(HighFive_libdeps PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "BOOST_ALL_NO_LIB;H5_USE_BOOST"
  INTERFACE_INCLUDE_DIRECTORIES "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/hdf5-1.12.2-ohajsg/include;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/boost-1.80.0-dcp2kq/include;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_externals/install_gcc-12.2.0-skylake/hpe-mpi-2.25.hmpt-mgws7w/include"
  INTERFACE_LINK_LIBRARIES "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/hdf5-1.12.2-ohajsg/lib/libhdf5.so;/usr/lib64/libpthread.so;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_externals/install_gcc-12.2.0-skylake/hpe-mpi-2.25.hmpt-mgws7w/lib/libmpi.so;/lib64/libpthread.so;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_externals/install_gcc-12.2.0-skylake/hpe-mpi-2.25.hmpt-mgws7w/lib/libmpi.so"
  INTERFACE_LINK_OPTIONS "SHELL:"
  INTERFACE_SYSTEM_INCLUDE_DIRECTORIES "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/hdf5-1.12.2-ohajsg/include;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_applications/install_gcc-12.2.0-skylake/boost-1.80.0-dcp2kq/include;/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_externals/install_gcc-12.2.0-skylake/hpe-mpi-2.25.hmpt-mgws7w/include"
)
```
Which is correct. The `-lhdf5-shared` from your output seems to point at something going wrong with HDF5's CMake finding/using/whatever

**Review by matz-e — APPROVED** [review]

**Review by pramodk — APPROVED** [review]
With my limited understating: with respect to `HIGHFIVE_USE_INSTALL_DEPS:Bool=Off`, users on BB5 deployment won't be affected I guess as most of software usage is via Spack. So this won't break anything I guess.

**Comment by matz-e** [comment]
I'm going to merge with the understanding that we are going to rework HighFive's CMake code to be more sane, anyways.

**Comment by pramodk** [comment]
As the viz team asked a few days ago, just want to tag @NadirRoGue to see if the updated deployment of brion/brayns is working now.

**Comment by NadirRoGue** [comment]
Thanks for tagging @pramodk 

The issue we have with HighFive comes from the way it is used on libsonata, which is generating a segfault at the momment. The changes on this PR do not affect that bug.

**Comment by pramodk** [comment]
@NadirRoGue - Stupid me! I was still thinking of our internal discussion to release a fix with a new high-five version and I didn't remember that this Spack PR doesn't do that. Sorry for the noise!

**Comment by NadirRoGue** [comment]
Not noise at all! Always good to keep up to date!

**Pull request merged**